examples/notebooks/qa-maximal-marginal-relevance.ipynb

```diff
@@ -0,0 +1,690 @@
+   "metadata": {},
+   "source": [
+    "First prepare a connection to a vector-search-capable Cassandra and initialize the required LLM and embeddings:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "id": "7672da56",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain.indexes.vectorstore import VectorStoreIndexWrapper\n",
+    "from langchain.vectorstores import Cassandra"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4578a87b",
+   "metadata": {},
+   "source": [
+    "A database connection is needed. _(If on a Colab, the only supported option is the cloud service Astra DB.)_"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "9515d4af-9fc2-4c29-b9f2-5c90cd48cc1b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Ensure loading of database credentials into environment variables:\n",
+    "import os\n",
+    "from dotenv import load_dotenv\n",
+    "load_dotenv(\"../../../.env\")\n",
+    "\n",
+    "import cassio"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "25ee44a4-8d6a-4b41-8a8f-7c93eed35034",
+   "metadata": {},
+   "source": [
+    "Select your choice of database by editing this cell, if needed:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "bfe78b92-7194-4ee9-ba7f-cd308409212c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "database_mode = \"cassandra\"  # \"cassandra\" / \"astra_db\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "fb5597b3-a028-4d8f-ad6d-8eaf0c941dcc",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "if database_mode == \"astra_db\":\n",
+    "    cassio.init(\n",
+    "        database_id=os.environ[\"ASTRA_DB_ID\"],\n",
+    "        token=os.environ[\"ASTRA_DB_APPLICATION_TOKEN\"],\n",
+    "        keyspace=os.environ.get(\"ASTRA_DB_KEYSPACE\"),  # this is optional\n",
+    "    )"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "329d969b-6532-4dc4-a315-1a5438a59919",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "if database_mode == \"cassandra\":\n",
+    "    from cqlsession import getCassandraCQLSession, getCassandraCQLKeyspace\n",
+    "    cassio.init(\n",
+    "        session=getCassandraCQLSession(),\n",
+    "        keyspace=getCassandraCQLKeyspace(),\n",
+    "    )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "50df3f86",
+   "metadata": {},
+   "source": [
+    "Below is the logic to instantiate the LLM and embeddings of choice. We chose to leave it in the notebooks for clarity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "e63d3c2e",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "LLM+embeddings from OpenAI\n"
+     ]
+    }
+   ],
+   "source": [
+    "import os\n",
+    "from llm_choice import suggestLLMProvider\n",
```

Comment:
This is machinery that in the original was kept in a separate module (which I don't see here?)
Also it is quite daunting (I sort of regret having given too much weight to "let's support various LLMs/embeddings"). I see it's been simplified to a vanilla OpenAI LLM instantiation (one line!) in the adoc.

1. could the same simplification occur here, perhaps?
2. I suspect this would not work as is since there's no "llm_choice.py". One may be better off with the "colab-ready" version, which through black magic is derived from this one in a way that can run self-containedly: https://github.com/CassioML/cassio-website/blob/main/docs/frameworks/langchain/.colab/colab_qa-maximal-marginal-relevance.ipynb

Reply:
Come to think of it, the "open in colab" link at the top of [this page](https://cassio.org/frameworks/langchain/qa-maximal-marginal-relevance/) is good as is - except probably you'll want to replace the langchain deps with ragstack

Reply:
1. I will apply these changes to this notebook in another PR. 
2. I added the llm_choice and cqlsession dependencies, and it passes CI so I'm merging. (Notebook doesn't have nbmake points to fail CI). I can revert to the other link if it doesn't work.